omp/test/solver/trs_kernels.cpp

```diff
@@ -0,0 +1,158 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/trs.hpp>
+
+
+#include <gtest/gtest.h>
+
+
+#include <random>
+
+
+#include <core/solver/trs_kernels.hpp>
+#include <core/test/utils.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+#include <ginkgo/core/stop/residual_norm_reduction.hpp>
+
+namespace {
+
+
+class Trs : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Dense<>;
+    using CsrMtx = gko::matrix::Csr<double, gko::int32>;
+    Trs() : rand_engine(30) {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        omp = gko::OmpExecutor::create();
+    }
+
+    void TearDown()
+    {
+        if (omp != nullptr) {
+            ASSERT_NO_THROW(omp->synchronize());
+        }
+    }
+
+    std::unique_ptr<Mtx> gen_mtx(int num_rows, int num_cols)
+    {
+        return gko::test::generate_random_lower_triangular_matrix<Mtx>(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(num_cols, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine, ref);
+    }
+
+    void initialize_data()
+    {
+        int m = 59;
+        int n = 43;
+        b = gen_mtx(m, n);
+        x = gen_mtx(m, n);
+        d_b = Mtx::create(omp);
+        d_b->copy_from(b.get());
+        d_x = Mtx::create(omp);
+        d_x->copy_from(x.get());
+        mat = gen_mtx(m, m);
+        csr_mat = CsrMtx::create(ref);
+        gko::as<gko::ConvertibleTo<CsrMtx>>(mat.get())->convert_to(
+            csr_mat.get());
+        d_mat = Mtx::create(omp);
+        d_mat->copy_from(mat.get());
+        d_csr_mat = CsrMtx::create(omp);
+        d_csr_mat->copy_from(csr_mat.get());
```

Comment:
Do we always do all of this for OpenMP? Technically, there's no need for `d_b` etc since `b` and so on are already in the correct place.

Reply:
Yeah, I think this is mainly for uniformity to show it between CUDA and OMP and anyother future executor. I would prefer to keep it like this, unless you have a reason to change it.

Reply:
Yes, we always do this and I would like to keep it that way. As it is right now, every data exists twice, and we can be sure that each algorithm computes on the same data. Otherwise, it could be possible that something you expect to not be modified is modified by one algorithm, and then reused in the other.
I know it is slower that way, but in my opinion, it adds a bit of safety.